man/run.1
.TH "RUNFILE\.SH" "1" "September 2026"
.SH "NAME"
\fBrunfile.sh – Language-agnostic project task runner built on the ubiquitous Make.\fR
.SH DESCRIPTION\\nBasic usage:
.RS 2
.nf
$ run \-\-help

· Usage · run ··················· Print list of all available tasks\.
          run [task] ············ Run a task\.
          run [task] [options] ·· Run a task with options available to task code\.

          run \-\-[help|version|runfile|makefile|edit|new|alias] [options]
          run \-[h|v|r|m|e|n|a] ·· Actions for managing Runfiles; details below\.

# \./Runfile \-\- syntax primer 
taskabc: # Task description\.
  shell command for task abc
taskxyz: taskabc # Task description, taskxyz runs taskabc first just like Make would\.
  shell command for task xyz
#^ Unlike Make, whitespace doesn't matter; tabs, spaces, extra blank lines are all ok\.

\-h \-\-help ··········· Print this usage documentation then exit\.
\-v \-\-version ········ Print current runfile\.sh version then exit\.
\-r \-\-runfile ········ Print contents of nearest Runfile \.
\-m \-\-makefile ······· Print contents of Makefile generated from nearest Runfile\.
\-m \-\-makefile TASK ·· Print contents of single task from generated Makefile\.
\-e \-\-edit ··········· Open nearest Runfile with \\$EDITOR\.
\-n \-\-new ············ Interactively create new Runfile in current dir\.
\-a \-\-alias ·········· Show command aliases for nearest Runfile \.
\-a \-\-alias FILENAME · Attempt to write/update aliases within shell config file\.
\-\-eject ············· Generate Makefile from Runfile and write to current dir\.

\-\-runfile\-verbose ··· Print code line\-by\-line to terminal during task execution\.
\-\-runfile\-compact ··· Use "compact" formatting for Runfile when creating or printing\.
\-\-runfile\-compat ···· Disable all features not compatible with Make\.
\-\-runfile\-confirm ··· Always ask for confirmation before opening files with \\$EDITOR\.
\-\-runfile\-noconfirm · Never ask for confirmation before opening files with \\$EDITOR\.
\-\-runfile\-noedit ···· Never open files with \\$EDITOR\.
RUNFILE_VERBOSE=1 RUNFILE_COMPACT=1   RUNFILE_COMPAT=1 ·  All options may also be  ·
RUNFILE_CONFIRM=1 RUNFILE_NOCONFIRM=1 RUNFILE_NOEDIT=1 · provided as env variables ·

\-\-make\-dry\-run ······ Don't execute task code, just print line\-by\-line to terminal\.
\-\-make\-ARGUMENT ····· Pass any argument directly to they underlying Make command
                    · by prefixing the intended Make argument with "\-\-make\-"\.
                    · For example, \-\-make\-dry\-run will pass \-\-dry\-run to Make\.
.fi
.RE
.SH Install
.P
Homebrew:
.RS 2
.nf
brew tap evnp/runfile\.sh && brew install runfile\.sh
.fi
.RE
.P
NPM:
.RS 2
.nf
npm install \-g runfile\.sh
.fi
.RE
.P
Curl:
.RS 2
.nf
read \-rp $'\\n'"Current directories in \\$PATH:"$'\\n'"$(echo $PATH|sed 's/:/\\n/g'|sort)"$'\\n\\n'"Enter a directory from the list above: " && [[ \-z "${REPLY}" ]] && echo "Cancelled " || ( curl \-L \-o "${REPLY/\\~/$HOME}/runfile\.sh" https://github\.com/evnp/runfile\.sh/raw/main/runfile\.sh && chmod +x "${REPLY/\\~/$HOME}/runfile\.sh" )
.fi
.RE
.P
runfile\.sh has no external dependencies, but it's good practice to audit code before downrunfile\.sh onto your system to ensure it contains nothing unexpected\. Please view the full source code for runfile\.sh here: https://github.com/evnp/runfile.sh/blob/master/runfile.sh
.P
If you also want to install runfile\.sh's man page:
.RS 2
.nf
read \-rp $'\\n'"Current directories in \\$:"$'\\n'"$(manpath|sed 's/:/\\n/g'|sort)"$'\\n\\n'"Enter a directory from the list above: " && [[ \-z "${REPLY}" ]] && echo "Cancelled " || curl \-L \-o "${REPLY/\\~/$HOME}/man1/runfile\.sh\.1" https://github\.com/evnp/runfile\.sh/raw/main/man/runfile\.sh\.1
.fi
.RE
.P
Verify installation:
.RS 2
.nf
runfile\.sh \-v
==> runfile\.sh 2\.0\.2

brew test runfile\.sh
==> Testing runfile\.sh
==> /opt/homebrew/Cellar/runfile\.sh/2\.0\.2/bin/runfile\.sh test \-\-print 1234 hello world
.fi
.RE
.SH Tests
.P
Run once:
.RS 2
.nf
npm install
npm test
.fi
.RE
.P
Use \fBfswatch\fP to re\-run tests on file changes:
.RS 2
.nf
brew install fswatch
npm install
npm run testw
.fi
.RE
.P
Non\-OSX: replace \fBbrew install fswatch\fP with package manager of choice 



run.sh
#!/usr/bin/env bash

# run :: v0.0.1

function create-runfile() {
	if [[ " $* " != *' --overwrite-runfile '* ]] && [[ -e 'Runfile' ]]
	then
		echo 'Runfile already exists. To overwrite, use:'
		echo 'run --overwrite-runfile'
		exit 1
	fi

	if [[ " $* " == *' --compact '* ]]
	then
cat <<EOF > Runfile
s start: end # start app
echo "start app"
run frontend
e end: # stop app
echo "stop app"
f frontend: # open frontend app
run start
echo "open frontend app"
b backend: # attach to backend server
run start
echo "attach to backend server"
r repl: # start shell in project environment
echo "start shell environment"
EOF
	else
cat <<EOF > Runfile
s start: end # start app
	echo "start app"
	run frontend

e end: # stop app
	echo "stop app"

f frontend: # open frontend app
	echo "open frontend app"

b backend: # attach to backend server
	echo "attach to backend server"

r repl: # start shell
	echo "start shell"
EOF
	fi
}

function lowercase-file() {
	echo "$1" | tr '[:upper:]' '[:lower:]'
}

function uppercase-file() {
	echo "$1" | tr '[:lower:]' '[:upper:]'
}

function titlecase-file() {
	echo "$( echo "$1" | cut -c1 | tr '[:lower:]' '[:upper:]' )$( echo "$1" | cut -c2- )"
}

function smartcase-file() { local name=''
	name="$( titlecase-file "$1" )"
	! [[ -e "${name}" && ! -d "${name}" ]] && name="$( lowercase-file "${name}" )"
	! [[ -e "${name}" && ! -d "${name}" ]] && name="$( uppercase-file "${name}" )"
	echo "${name}"
}

function edit-file-smartcase() { local name=''
	name="$( smartcase-file "$1" )"
	if [[ " $* " != *' --noedit '* ]]
	then
		[[ " $* " == *' --confirm '* && " $* " != *' --noconfirm '* ]] && \
			read -rsn1 -p "Press any key to edit ${name} with $EDITOR · CTRL+C to exit"
		$EDITOR "${name}"
	fi
}

function print-file-smartcase() {
	cat "$( smartcase-file "$1" )"
}

function print-makefile() {
	sed -E "s!\t${at}make --makefile ${mf} !\t${at}make !" "$@"
}

function cd-to-nearest-file() { local lower='' upper='' title=''
	lower="$( lowercase-file "$1" )"
	upper="$( uppercase-file "$1" )"
	title="$( titlecase-file "$1" )"
	while [[ -d "./${lower}" || -d "./${upper}" || -d "./${title}" ]] || \
			! [[ -e "./${lower}" || -e "./${upper}" || -e "./${title}" ]]
	do
		if [[ "$( pwd )" != "$HOME" ]]
		then
			cd ..
		else
			echo "No ${title} found. To create one here, use:"
			echo "run --create-${lower}"
			# Note: This message pertains to the user's shell which _won't_ have
			# changed directory, because this script's main function uses a subshell.
			# So there's no need to change directory back to where we started here.
			exit 1
		fi
	done
}

function main() ( set -euo pipefail
	local mf='' at='' ws='' arg='' args=() cmd="" rewrite=""

	# Handle various optional actions:
	[[ " $* " == *' --create-runfile '* || " $* " == *' --overwrite-runfile '* ]] && \
		create-runfile "$@" && edit-file-smartcase runfile --confirm "$@" && exit 0
	[[ " $* " == *' --print-runfile '* ]] && \
		cd-to-nearest-file runfile && print-file-smartcase runfile && exit 0
	[[ " $* " == *' --edit-runfile '* ]] && \
		cd-to-nearest-file runfile && edit-file-smartcase runfile "$@" && exit 0
	[[ " $* " == *' --edit-makefile '* ]] && \
		cd-to-nearest-file makefile && edit-file-smartcase makefile "$@" && exit 0

	# If no runfile in current dir, navigate up looking for one until we reach $HOME:
	cd-to-nearest-file runfile

	# Local values:
	mf="$( mktemp )"	# Temporary makefile which we will pass to make.
	at="@"						# @-prefix causes make to execute commands silently.
	args=()						# Arguments that will be passed on to invoked run command.

	# Existing Makefile Compatibility:
	# If 'run cmd' or 'make cmd' appears within another command in a Runfile,
	# normally we'd rewrite these both to 'make --makefile ${mf}' to avoid invoking
	# runfile.sh recursively. However, if a separate Makefile exists in the same directory
	# as the current Runfile we're executing a command for, we WON'T rewrite 'make cmd'
	# in this way. In that case, leaving 'make cmd' alone allows the user to reference
	# a command in the existing Makefile from their runfile.sh command.
	rewrite="(make|run)"
	if [[ -e 'Makefile' && ! -d 'Makefile' ]] \
	|| [[ -e 'makefile' && ! -d 'makefile' ]] \
	|| [[ -e 'MAKEFILE' && ! -d 'MAKEFILE' ]]
	then
		rewrite="run"
	fi

	# Handle these args which we don't want to pass on to command:
	# --print-command, --dry-run-command,
	# --overwrite-runfile, --overwrite-makefile,
	# --create-runfile, --create-makefile,
	# --print-runfile, --print-makefile,
	# --edit-runfile, --edit-makefile,
	for arg in "$@"
	do
		if [[ "${arg}" == '--print-command' || "${arg}" == '--dry-run-command' ]]
		then
			at="" # Remove @-prefix so that make prints commands before executing them.
		elif ! [[ "${arg}" == '--overwrite-runfile' || "${arg}" == '--overwrite-makefile' ]] \
			&& ! [[ "${arg}" == '--create-runfile' || "${arg}" == '--create-makefile' ]] \
			&& ! [[ "${arg}" == '--print-runfile' || "${arg}" == '--print-makefile' ]] \
			&& ! [[ "${arg}" == '--edit-runfile' || "${arg}" == '--edit-makefile' ]]
		then
			if [[ -z "${cmd}" ]]
			then
				cmd="${arg}"
			else
				args+=( "${arg}" )
			fi
		fi
	done

ws="[[:space:]]"						# whitespace pattern
ch="[a-zA-Z0-9_-]"					# character pattern
wc="[[:space:]a-zA-Z0-9_-]"	# whitespace & character pattern

# ::::::::::::::::::::::::::::::::::::::::::
# Construct temporary Makefile from Runfile:
cat <<EOF > "${mf}"
help: .usage

$(
	sed -Ee "s!^[[:space:]]*!\t${at}!" \
			-e "s!^\t${at}(${wc}+${ch})${ws}*:(${wc}+${ch})?${ws}*#${ws}*(.*)\$!\1:\2 # \3!" \
			-e "s!^\t${at}${rewrite} !\t${at}make --makefile ${mf} !" \
			-e "s!\t${at}\$!\t!" \
		Runfile
)

.usage:
	@grep -E "^${wc}+:${wc}*# " \$(MAKEFILE_LIST) \\
	| sed -Ee 's/^/\\t/' -e "s/${ws}*:${wc}*#${ws}*/ · /"
EOF
# Done with temporary Makefile construction.
# ::::::::::::::::::::::::::::::::::::::::::

	# --create-makefile : Write generated Makefile then exit.
	# --overwrite-makefile : Can be used to overwrite when Makefile already exists.
	if [[ " $* " == *' --create-makefile '* || " $* " == *' --overwrite-makefile '* ]]
	then
		if [[ " $* " != *' --overwrite-makefile '* ]] \
		&& [[ -e 'Makefile' && ! -d 'Makefile' ]]
		then
			echo 'Makefile already exists. To overwrite, use:'
			echo 'make --overwrite-makefile'
			rm "${mf}"
			exit 1
		else
			print-makefile "${mf}" > ./Makefile
			rm "${mf}"
			exit 0
		fi
	fi

	# --print-makefile : Print generated Makefile then exit.
	if [[ " $* " == *' --print-makefile '* ]]
	then
		print-makefile "${mf}"
		rm "${mf}"
		exit 0
	fi

	# Main Path : Invoke make with generated makefile and all other arguments.
	if [[ " $* " == *' --dry-run-command '* ]]
	then
		make --makefile "${mf}" --dry-run "${cmd}"
	elif [[ -n "${cmd}" ]]
	then
		make --makefile "${mf}" "${cmd}"
	else
		make --makefile "${mf}"
	fi

	# Clean up temporary Makefile and exit with success:
	rm "${mf}"
	exit 0
)

main "$@"
